### PR TITLE
Allow local closure allocations

### PR DESCRIPTION
**Pull request by stedolan**
Add middle and back-end support to allow closures to be allocated locally.

This patch currently contains a bug (see the FIXME in cmmgen.ml): sometimes, it seems that the parts of a group of mutally recursive closures might not agree on local-ness. (I haven't debugged this fully but I think it might be to do with partial applications)

There are a few testsuite failures here:
  - the local types tests need updating
  - the systhreads tests are broken (needs better GC support, in a future patch)
  - a backtrace test breaks due to tail calls changing (a known bug, to be fixed in a future patch)
  - a test in typing-gadts fails (an upstream bug, fixed in ocaml/ocaml#10638)

### CODE REVIEW

File: lambda/translcore.ml [hunk 1]
```diff
@@ -748,22 +750,24 @@ and transl_apply ~scopes
...
             Lfunction{kind = Curried; params = ids; return;
-                      body = lam; attr; loc} ->
+                      body = lam; attr; loc; mode} ->
               Lfunction{kind = Curried;
                         params = (id_arg, Pgenval)::ids;
                         return;
                         body = lam; attr;
-                        loc}
+                        loc;
+                        mode}
```
Comment (lpw25) [inline lambda/translcore.ml]: Could we add a FIXME comment here too, because I think that re-using the mode for this function is not always correct.

**Review by lpw25 — APPROVED** [review]
The parts not marked "FIXME" look good to me.

**Comment by stedolan** [comment]
I've pushed a fix fixing the bug with `let rec` groups - now all definitions in a let rec share a mode variable. The testsuite still needs some work, though.

**Review by lpw25 — APPROVED** [review]

**Comment by stedolan** [comment]
I've fixed up the local types tests, so merging. (Remaining testsuite failures are for future patches)

**Pull request merged**